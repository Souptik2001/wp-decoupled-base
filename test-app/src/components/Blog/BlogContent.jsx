import { useMemo } from "react";

function BlogContent({blogData}) {

	const dateString = useMemo(() => {
		var dateObject = new Date(blogData.date);
		var year = dateObject.getFullYear();
		var date = dateObject.getDate();
		var month = dateObject.getMonth();
		return `${date}/${month}/${year}`;
	}, [blogData.date]);

	return (
		<div>
			<h1>{blogData.title.rendered}</h1>
			<em>Written by <strong>{blogData.author.name}</strong> on <strong>{dateString}</strong></em>
			<div
				dangerouslySetInnerHTML={{__html: blogData.content.rendered}}
			/>
		</div>
	)

}

export default BlogContent;