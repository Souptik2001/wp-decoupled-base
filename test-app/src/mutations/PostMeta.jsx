import { gql } from "@apollo/client";

export const UPDATE_POST_META = gql`
	mutation UpdatePostMeta($postType: String!, $postID: Int!, $metaInput: [MetaInput!]!) {
		updatePostMeta(postType: $postType, postID: $postID, metaInput: $metaInput) {
			success
		}
	}
`;
