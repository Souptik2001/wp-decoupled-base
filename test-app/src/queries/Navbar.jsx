import { gql } from "@apollo/client";

export const GET_NAVBAR = gql`
	query GetMenus($slug: String) {
		menu(slug: $slug) {
			slug
			name
			term_id
			items {
				ID
				menu_order
				menu_item_parent
				title
				url
				type
				object
				child_items {
					ID
					menu_order
					menu_item_parent
					title
					url
					type
					object
				}
			}
		}
	}
`;
