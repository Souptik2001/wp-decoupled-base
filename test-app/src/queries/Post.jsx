import { gql } from "@apollo/client";

export const GET_POST = gql`
	query GetPost($postType: String!, $slug: String!) {
		post(postType: $postType, slug: $slug) {
			title {
				rendered
			}
			slug
			id
			date
			content {
				rendered
			}
			author {
				id
				name
				slug
				link
			}
			meta {
				post_views
			}
		}
	}
`;
