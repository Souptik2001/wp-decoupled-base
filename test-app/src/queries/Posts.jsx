import { gql } from "@apollo/client";

export const GET_POSTS = gql`
	query GetPosts($limit: Int, $postType: String!, $offset: Int!, $taxonomy: String, $term: Int) {
		posts(limit: $limit, postType: $postType, offset: $offset, taxonomy: $taxonomy, term: $term) {
			pageInfo {
				endCursor
				hasNextPage
				totalPages
			}
			posts {
				type
				excerpt {
					rendered
				}
				date
				slug
				title {
					rendered
				}
			}
		}
	}
`;