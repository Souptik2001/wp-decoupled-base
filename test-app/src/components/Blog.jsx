import { useQuery } from "@apollo/client";
import { useContext, useEffect } from "react";
import { useParams } from "react-router-dom";
import { GET_POST } from "../queries/Post";
import BlogContent from "./Blog/BlogContent";
import { PostStoreContext } from "./BlogListRoutes";
import PageNotFound from "./PageNotFound";

function Blog() {
	const {slug} = useParams();
	const {postTypeData} = useContext(PostStoreContext);
	const {data, loading} = useQuery(
		GET_POST,
		{
			variables: {
				slug,
				postType: postTypeData?.rest_base
			}
		}
	);

	useEffect(() => {
		if (data && data.post) {
			document.title = data.post.title.rendered;
		}
	}, [data]);

	return (
		<div>
			{
				loading
				&&
				(<h1>Loading...</h1>)
			}
			{
				! loading
				&&
				data?.post
				&&
				(<BlogContent blogData={data?.post} />)

			}
			{
				! loading
				&&
				! data?.post
				&&
				(<PageNotFound />)
			}
		</div>
	);
}

export default Blog;