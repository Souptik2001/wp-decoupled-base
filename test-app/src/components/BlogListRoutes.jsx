import { useQuery } from "@apollo/client";
import { createContext, useReducer } from "react";
import { Route, Routes, useParams } from "react-router-dom";
import { GET_POST_TYPE } from "../queries/PostType";
import Blog from "./Blog/Blog";
import BlogList from "./BlogList/BlogList";
import Filters from "./BlogList/Filters";
import PageNotFound from "./PageNotFound";

export const PostStoreContext = createContext();

function updatePosts(state, action) {
	const {taxonomy, term, postType, page, total, termID} = action;

	if ( ! state[postType] ) {
		state[postType] = {};
	}

	if ( taxonomy && ! state[postType][taxonomy] ) {
		state[postType][taxonomy] = {};
	}

	if ( taxonomy && term ) {
		if ( ! state[postType][taxonomy][term] ) {
			state[postType][taxonomy][term] = {};
		}
		state[postType][taxonomy][term]['currentPage'] = page ? page : state[postType][taxonomy][term]['currentPage'];
		state[postType][taxonomy][term]['totalPages'] = total ? total : state[postType][taxonomy][term]['totalPages'];
		state[postType][taxonomy][term]['id'] = termID ? termID : state[postType][taxonomy][term]['id'];
	} else {
		state[postType]['no filter'] = {
			currentPage: page,
			totalPages: total
		}
	}

	return {...state};
}

function BlogListRoutes() {

	const {postType=""} = useParams();

	const [posts, setPosts] = useReducer(updatePosts, {});

	const {data, loading} = useQuery(
		GET_POST_TYPE,
		{
			variables: {
				type: postType
			}
		}
	);

	if (loading) {
		return(
			<h1>Loading...</h1>
		)
	}

	const postStore = {
		posts,
		setPosts,
		postTypeData: data?.postType
	};

	return (
		<div>
			{
				! loading
				&&
				data?.postType
				&&
				<PostStoreContext.Provider value={postStore}>
					<Routes>
						<Route element={<Filters />}>
							<Route path=":page?" element={<BlogList />}  />
							<Route path='filter/:taxonomy/:term/:page?' element={<BlogList />} />
							<Route path=':year/:month/:date/:slug' element={<Blog />} />
						</Route>
					</Routes>
				</PostStoreContext.Provider>
			}
			{
				! loading
				&&
				! data?.postType
				&&
				(<PageNotFound />)
			}
		</div>
	)

}

export default BlogListRoutes;
