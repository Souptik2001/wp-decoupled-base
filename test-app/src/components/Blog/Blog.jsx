import { useQuery } from "@apollo/client";
import { useContext, useEffect } from "react";
import { useParams } from "react-router-dom";
import { GET_POST } from "../../queries/Post";
import { PostStoreContext } from "../BlogListRoutes";
import Loading from "../Loading";
import PageNotFound from "../PageNotFound";
import BlogContent from "./BlogContent";

function Blog() {
	const {slug} = useParams();
	const {postTypeData} = useContext(PostStoreContext);
	const {data, loading} = useQuery(
		GET_POST,
		{
			variables: {
				slug,
				postType: postTypeData?.rest_base
			}
		}
	);

	useEffect(() => {
		if (data && data.post) {
			document.title = data.post.title.rendered;
		}
	}, [data]);

	return (
		<div data-testid="single-post">
			{
				loading
				&&
				(<Loading />)
			}
			{
				! loading
				&&
				data?.post
				&&
				(<BlogContent blogData={data?.post} />)
			}
			{
				! loading
				&&
				! data?.post
				&&
				(<PageNotFound />)
			}
		</div>
	);
}

export default Blog;