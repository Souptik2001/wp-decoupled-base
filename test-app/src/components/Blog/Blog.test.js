import { ApolloClient, ApolloProvider, InMemoryCache } from '@apollo/client';
import { MockedProvider } from '@apollo/client/testing';
import { render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { Route, Routes } from 'react-router-dom';
import { StaticRouter } from 'react-router-dom/server';
import { UPDATE_POST_META } from '../../mutations/PostMeta';
import { GET_POST } from '../../queries/Post';
import { PostStoreContext } from '../BlogListRoutes';
import Blog from './Blog';

// TODO: Mock data is not being received. The original is returned - Pass the same variables in the mockdata structure. MockedProvider component should be inside ApolloProvider.
// TODO: Confirm that which variable to use in snapshot.
// TODO: Router not working.

const client = new ApolloClient({
  uri: process.env.REACT_APP_GRAPHQL_URL,
  cache: new InMemoryCache()
});

const postStore = {
	postTypeData: {
		rest_base: "blogs"
	}
};

test('renders single blog page', async () => {
	const initialViews = 10;
	const increasedView = initialViews + 1;
	const blogDataMock = {
		request: {
			query: GET_POST,
			variables: {
				slug: 'test',
				postType: 'blogs'
			}
		},
		result: {
			"data": {
				"post": {
				  "title": {
					"rendered": "Mock Test"
				  },
				  "slug": "test",
				  "id": 5,
				  "date": "2023-04-14T15:45:57",
				  "content": {
					"rendered": "\n<p>This is a mock test data.</p>\n"
				  },
				  "author": {
					"id": 1,
					"name": "admin",
					"slug": "admin",
					"link": "https://test.lndo.site/author/admin/"
				  },
				  "meta": {
					"post_views": initialViews
				  }
				}
			}
		}
	};

	// This is for the mutation called for updating the post count.
	const blogUpdateMockData = {
		request: {
			query: UPDATE_POST_META,
			variables: {
				postType: "blogs",
				postID: 5,
				metaInput: [
					{
						meta_key: "post_views",
						meta_value: increasedView
					}
				]
			}
		},
		result: {
			"data": {
			  "updatePostMeta": {
				"success": true,
				"data": {
				  "id": 5,
				  "title": {
					"rendered": "Mock Test"
				  },
				  "author": {
					"id": 1,
					"link": "https://test.lndo.site/author/admin/",
					"name": "admin",
					"slug": "admin"
				  },
				  "content": {
					"rendered": "\n<p>This is a mock test data.</p>\n",
					"protected": false
				  },
				  "date": "2023-04-14T15:45:57",
				  "slug": "test",
				  "type": "blog",
				  "excerpt": {
					"rendered": "<p>This is a test.</p>\n"
				  },
				  "meta": {
					"post_views": increasedView
				  }
				}
			  }
			}
		}
	}

	const route = "/blog/2023/3/14/test";

	const {baseElement} = render(
		<StaticRouter location={route}>
			<ApolloProvider client={client}>
				<React.StrictMode>
					{
						/** Have to provide blogUpdateMockData twice, so that it doesn't throw warning that "No mock response found."
						 * Not exactly sure why. But in a GH discussion found that we have to provide the mock query and response as many times, that the query is actually called.
						 * So, because of the same development mode issue the mutation is called twice, and because of that the mock should be provided twice.
						 */
					}
					<MockedProvider addTypename={false} mocks={[blogDataMock, blogUpdateMockData, blogUpdateMockData]} >
						<PostStoreContext.Provider value={postStore}>
							<Routes>
								<Route path='/blog/:year/:month/:date/:slug' element={<Blog />} />
							</Routes>
						</PostStoreContext.Provider>
					</MockedProvider>
				</React.StrictMode>
			</ApolloProvider>
		</StaticRouter>
	);

	await waitFor(
		() => screen.findByTestId("single-post-content"),
		{
			timeout: 1000 * 60
		}
	);

	expect(baseElement.outerHTML).toMatchSnapshot();
});
