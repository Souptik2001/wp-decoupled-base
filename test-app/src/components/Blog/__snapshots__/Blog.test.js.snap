// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders single blog page 1`] = `
"<body><div><div data-testid=\\"single-post\\"><div data-testid=\\"single-post-content\\"><h1>Mock Test</h1><em>Written by <strong>admin</strong> on <strong>14/3/2023</strong></em><br><em>Views: <strong>10</strong></em><div>
<p>This is a mock test data.</p>
</div></div></div></div></body>"
`;
